src/metabase/api/card.clj

```diff
@@ -280,24 +267,58 @@
                                     :dataset (boolean (:dataset card-data))
                                     :parameters (or parameters [])
                                     :parameter_mappings (or parameter_mappings []))
-        result-metadata-chan (result-metadata-async {:query dataset_query
+        result-metadata-chan (result-metadata-async {:query    dataset_query
                                                      :metadata result_metadata
                                                      :dataset? dataset})
-        out-chan             (a/promise-chan)]
-    (a/go
-      (try
-        (let [card-data (assoc card-data :result_metadata (a/<! result-metadata-chan))]
-          (a/close! result-metadata-chan)
-          ;; now do the actual saving on a separate thread so we don't tie up our precious core.async thread. Pipe the
-          ;; result into `out-chan`.
-          (async.u/promise-pipe (save-new-card-async! card-data @api/*current-user*) out-chan))
-        (catch Throwable e
-          (a/put! out-chan e)
-          (a/close! out-chan))))
-    ;; Return a channel
-    out-chan))
-
-(api/defendpoint ^:returns-chan POST "/"
+        metadata-timeout     (a/timeout metadata-sync-wait-ms)
+        [metadata port]      (a/alts!! [result-metadata-chan metadata-timeout])
+        timed-out?           (= port metadata-timeout)
+        card                 (db/transaction
+                              ;; Adding a new card at `collection_position` could cause other cards in this
+                              ;; collection to change position, check that and fix it if needed
+                              (api/maybe-reconcile-collection-position! card-data)
+                              (db/insert! Card (cond-> card-data
+                                                 (not timed-out?)
+                                                 (assoc :result_metadata metadata))))]
+    (events/publish-event! :card-create card)
+    (log/info (trs "Metadata not available soon enough. Saving new card and asynchronously updating metadata"))
+    ;; include same information returned by GET /api/card/:id since frontend replaces the Card it currently has with
+    ;; returned one -- See #4283
+    (u/prog1 (-> card
+                 (hydrate :creator
+                          :dashboard_count
+                          :can_write
+                          :average_query_time
+                          :last_query_start
+                          :collection [:moderation_reviews :moderator_details])
+                 (assoc :last-edit-info (last-edit/edit-information-for-user @api/*current-user*)))
+      (when timed-out?
+        (.submit clojure.lang.Agent/pooledExecutor
```

Comment:
I'm not convinced `Agent/pooledExecutor` is the best executor to use. Its size is nCPUs + 2, so on a small EC2 instance it's 3 or 4. OTOH it's 34 on my home computer.

Neither one really matches up to the IRL effective limit of 15 open connections per data warehouse DB. 

My only concern is that on instances with a lot of users we could be waiting a while if a lot of people create Cards at the same time.

Since the QP already handles thread pool nonsense and connection pooling is handled by the drivers maybe just doing this with core async would be better. We already have `result-metadata-chan` anyway

Reply:
Good point. My main consideration here was to make sure we didn't run into shenanigans with toucan's connection closing. I can switch this around. Maybe I can finish waiting in a `go` block and then just submit somewhere (or even a future?) to actually save the the results to the db. That execution will need some thread somewhere but it its only concerned with saving (and not waiting on the channel) it should be fine. Good point.

Reply:
👍 I/O should go on a separate thread so we don't tie up our precious limited core.async threads (I think the pool is fixed at like 4 or 8 threads or something) but `future` should be fine there.

I guess ideally maybe you want to do the `db/select` on a separate thread too for the same reason (you can use `a/thread` to run it on a different thread and get a channel to get the result when it comes back).